firefox-ios/firefox-ios-tests/Tests/ClientTests/Frontend/Homepage Rebuild/Redux/TopSitesMiddlewareTests.swift

```diff
@@ -0,0 +1,98 @@
+    var appState: AppState!
+
+    override func setUp() {
+        super.setUp()
+        DependencyHelperMock().bootstrapDependencies()
+        setupStore()
+    }
+
+    override func tearDown() {
+        DependencyHelperMock().reset()
+        resetStore()
+        super.tearDown()
+    }
+
+    func test_homepageInitializeAction_returnsTopSitesSection() throws {
+        let subject = createSubject(topSitesManager: mockTopSitesManager)
+        let action = HomepageAction(windowUUID: .XCTestDefaultUUID, actionType: HomepageActionType.initialize)
+
+        let expectation = XCTestExpectation(description: "All relevant top sites middleware actions are dispatched")
+
+        mockStore.waitForDispatch(3) {
+            expectation.fulfill()
+        }
+
+        subject.topSitesProvider(appState, action)
+
+        wait(for: [expectation])
+
+        XCTAssertEqual(mockTopSitesManager.getOtherSitesCalledCount, 1)
+        XCTAssertEqual(mockTopSitesManager.fetchSponsoredSitesCalledCount, 1)
+        XCTAssertEqual(mockTopSitesManager.recalculateTopSitesCalledCount, 3)
+
+        let actionsCalled = try XCTUnwrap(mockStore.dispatchedActions as? [TopSitesAction])
+        let actionsType = try XCTUnwrap(actionsCalled.compactMap { $0.actionType } as? [TopSitesMiddlewareActionType])
+
+        XCTAssertEqual(mockStore.dispatchedActions.count, 3)
+        XCTAssertEqual(actionsType, [.retrievedUpdatedSites, .retrievedUpdatedSites, .retrievedUpdatedSites])
+        XCTAssertEqual(actionsCalled.last?.topSites?.count, 30)
+    }
+
+    func test_fetchTopSitesAction_returnsTopSitesSection() throws {
+        let subject = createSubject(topSitesManager: mockTopSitesManager)
+        let action = TopSitesAction(windowUUID: .XCTestDefaultUUID, actionType: TopSitesActionType.fetchTopSites)
+
+        let expectation = XCTestExpectation(description: "All top sites middleware actions are dispatched")
+
+        mockStore.waitForDispatch(3) {
+            expectation.fulfill()
+        }
+
+        subject.topSitesProvider(appState, action)
+
+        wait(for: [expectation])
+
+        XCTAssertEqual(mockTopSitesManager.getOtherSitesCalledCount, 1)
+        XCTAssertEqual(mockTopSitesManager.fetchSponsoredSitesCalledCount, 1)
+        XCTAssertEqual(mockTopSitesManager.recalculateTopSitesCalledCount, 3)
+
+        let actionsCalled = try XCTUnwrap(mockStore.dispatchedActions as? [TopSitesAction])
+        let actionsType = try XCTUnwrap(actionsCalled.compactMap { $0.actionType } as? [TopSitesMiddlewareActionType])
+
+        XCTAssertEqual(mockStore.dispatchedActions.count, 3)
+        XCTAssertEqual(actionsType, [.retrievedUpdatedSites, .retrievedUpdatedSites, .retrievedUpdatedSites])
+        XCTAssertEqual(actionsCalled.last?.topSites?.count, 30)
+    }
+
+    // MARK: - Helpers
+    private func createSubject(topSitesManager: MockTopSitesManager) -> TopSitesMiddleware {
+        return TopSitesMiddleware(topSitesManager: topSitesManager)
+    }
```

Comment:
Could use a `trackForMemoryLeaks` here? 👀

Reply:
🤔 I realize we aren't using this in any of our other middlewares either, making this call causes tests to fail for our middleware test in general. I'm not sure if its because the middlewares are instantiate on the store level. Do you mind if I create a separate ticket to investigate to avoid blocking this PR?

Reply:
Might be due to that. I realized I should have asked, does our middlewares should be deinit at some point in our app lifecycle? If not maybe there's not a lot of value with this `trackForMemoryLeaks` in that case and it's possibly a non-issue